src/arch/helpers390x_128.h

```diff
@@ -0,0 +1,535 @@
+//          Copyright Naoki Shibata 2010 - 2019.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+
+#if CONFIG == 130 || CONFIG == 140 || CONFIG == 131 || CONFIG == 141
+
+#if !defined(__VX__)
+#error This helper is for IBM s390x.
+#endif
+
+#if (CONFIG == 140 || CONFIG == 141) && __ARCH__ < 12
+#error Please specify -march=z14.
+#endif
+
+#else
+#error CONFIG macro invalid or not defined
+#endif
+
+#define ENABLE_DP
+#define LOG2VECTLENDP 1
+#define VECTLENDP (1 << LOG2VECTLENDP)
+
+#define ENABLE_SP
+#define LOG2VECTLENSP (LOG2VECTLENDP+1)
+#define VECTLENSP (1 << LOG2VECTLENSP)
+
+#if CONFIG == 130 || CONFIG == 140
+#define ENABLE_FMA_DP
+#define ENABLE_FMA_SP
+#endif
+
+#define ACCURATE_SQRT
+#define FULL_FP_ROUNDING
+
+#include <vecintrin.h>
+
+#include <stdint.h>
+#include <math.h>
+#include "misc.h"
+
+#define VECTOR __attribute__((vector_size(16)))
+
+typedef VECTOR unsigned long long vmask;
+typedef VECTOR unsigned long long vopmask;
+
+typedef VECTOR double vdouble;
+typedef VECTOR int vint;
+
+typedef VECTOR float vfloat;
```

Comment:
Z13 doesn't support single-precision, but it can be emulated via two double-precision registers

Reply:
I don't understand this.
It seems that both gcc and qemu support single-precision vector operation with z13 option.
Are both of those buggy?

Reply:
Ah, gcc unrolls each vector operation into scalar operations.

Reply:
vector extension in both compilers(clang & gcc) always emulate non-existence instructions which may lead to bad performance, that why I prefer the use of the prototypes of `ZVECTOR` and `VSX` instead. 
An example of how we should handle single-precision in `Z13`:
```C

// type
#if CONFIG <= 131
typedef struct {
    vdouble val[2];
} vfloat;
#else
typedef __vector float vfloat;
#endif
// load
static INLINE vfloat vloadu_vf_p(const float *p)
{
#if CONFIG <= 131
    vfloat r;
    r.val[0] = vec_ld2f(p); // load and convert
    r.val[1] = vec_ld2f(p + 2);
    return r;
#else
	return vec_xl(0, p);
#endif
}
// store
static INLINE void vstoreu_v_p_vf(float *p, vfloat v)
{
#if CONFIG <= 131
    vec_st2f(v.val[0], p); // convert and store
    vec_st2f(v.val[1], p + 2);
#else
    return vec_xst(v, 0, p);
#endif
}
// Now emulate all operations via two double-prescion vectors
static INLINE vfloat vsqrt_vf_vf(vfloat vf)
{
#if CONFIG <= 131
    vf.val[0] = vec_sqrt(vf.val[0]);
    vf.val[1] = vec_sqrt(vf.val[1]);
    return vf;
#else
    return vec_sqrt(vf); 
#endif
}
````

Reply:
This is emulation using double-precision operations.
SLEEF has functions that return bit-identical results across all platforms, and those functions cannot be implemented using this method. We need genuine single-precision operations.
I don't know how widely Z13 computers are currently deployed, but is single-precision support for Z13 so important?

Reply:
Z13 mainframe launched in 2015, not sure if still widely used but most of the Linux-one instances provided through IBM/Cloud are Z14 so I guess we can drop the support of Z13 for now,
and focuses only on Z14/Z15